src/xmipp/libraries/reconstruction_starpu/mpi/mpi_reconstruct_fourier_starpu.h

```diff
@@ -0,0 +1,111 @@
+#include <mpi.h>
+#include <deque>
+
+/**
+ * MPI extension of ProgRecFourierStarPU.
+ * MPI should already be initialized externally. (This is done by runReconstructFourierMpiStarPU function)
+ *
+ * At least two program instances are needed - one master and at least one worker (though any real speedup is expected
+ * only with multiple workers). It is expected, but not required, that the master will run on the same machine as one
+ * of the workers, as it uses minimal processing power.
+ *
+ * All program instances, no matter their role, should have the access to identical input files (through network storage,
+ * manual copy, etc.).
+ *
+ * First worker (i.e. second global rank) will write out the result.
+ */
+class ProgRecFourierMpiStarPU : public ProgRecFourierStarPU {
+
+	bool mpiInitialized = false;
+
+	/** 0..1, how many batches should be distributed by default?
+	* Larger values lead to less overhead in homogeneous environment,
+	* while smaller values can distribute batches better in more heterogeneous environment.
+	* At least one batch is always given to workers, no matter what. */
+	float percentOfJobsDistributedByDefault = 0.3f;
+	/** Amount of batches that should always be in the pipeline, if possible.
+	* Similar tradeoff like percentOfJobsDistributedByDefault,
+	* but some number is always needed so that the pipeline does not stall. */
+	uint32_t preferredBatchesInPipeline = 5;
+
+public:
+
+	/** Specify supported command line arguments */
+	void defineParams() override;
+
+	/** Read arguments from command line */
+	void readParams() override;
+
+	/** Read with MPI Initialization */
+	void read(int argc, char **argv, bool reportErrors = true) override;
+
+	/** try with MPI abort */
+	int tryRun() override;
+
+	void run() override;
+
+	/** MPI shutdown */
+	~ProgRecFourierMpiStarPU() override;
+
+private:
+
+	/** How many and which batches should each worker get by default.
+	 * @param batchCount how many batches are there to distribute
+	 * @param workerCount between how many workers should the work be distributed
+	 * @param selfWorkerId if called from worker id, myBatches will be filled with the numbers of batches to be processed
+	 * by default by this worker. The worker can begin to work on these right away.
+	 * @param myBatches will be filled with numbers of batches on which this worker should work on, may be nullptr
+	 * @return amount of batches distributed by default, equal to the next batch to distribute. If this value is equal to batchCount,
+	 * it means that all batches have been distributed by default and dynamic batch distribution is not needed. */
+	uint32_t defaultBatchDistribution(const uint32_t batchCount, const uint32_t workerCount, const uint32_t selfWorkerId, std::deque<uint32_t>* myBatches) const;
+
+	/** Distributes the work to workers. Does not do any work itself.
+	 * @param workerCount amount of workers */
+	void runMaster(uint32_t workerCount);
```

Comment:
Same reason as above. But I can change it, if you want to.

Note that I don't use it as an optimization (it is probably very slightly slower, I don't know), but because it better fits the character of the value.

Reply:
Just make sure it's not used at the place where it can actually overflow. No need to refactor the whole thing

Reply:
I have changed the types of variables referring to MPI group sizes or ranks to `int`, since that is what MPI uses.